### PR TITLE
Identified dead_percentage and reanimating

### LINKED ISSUE #237: df.world_data.xml:world_region reanimation investigation
I've done some investigation regarding reanimation. Quietust commented that unk_1e4 and unk_1e6 used to be related to reanimation in old DF versions when the evil/good flags were proposed to be named, and I've used that as a starting assumption.

This issue is raised with the hope of starting an investigation discussion that can eventually lead to a structure update pull request.

As far as I've seen, a region interaction involving reanimation associated to an evil region does not actually result in the embark being reanimating, as such embarks frequently end up being non animating.
Generating a world with extra evil regions I've embarked a fair number of times, and my preliminary results are:
- unk_1e6 generated with a value of 256 causes the evil embark to be reanimating, while 0 causes it not to be.
- unk_1e6 has been seen to have a value of 1 on neutral regions only. Only 0, 1, and 256 has been observed.
- unk_1e6 DFHacked from 0 to 256 does NOT seem to work to convert an evil "reanimating" region into an actually reanimating one, so there is probably some additional requirement somewhere.
- unk_1e4 with a value of 100 caused every piece of greenery to be dead at the embark of an evil "reanimating" biome, and DFHacking 0 -> 100 caused a non dead region to be a 100% dead one on a subsequent embark (beside the previous fortress).
- unk_1e4 with a value of 59 had about half of the trees being dead.
- unk_1e4 has been observed to have a value > 256, but I haven't seen the first byte to have any bit other than bit 0 set.
- Actually reanimating biomes have been observed with unk_1e4 = 0, so unk_1e4 is probably not involved in that distinction.
- Both unk_1e4 and unk_1e6 are set during worldgen BEFORE good/evil is assigned. Thus, they can have non zero values for regions that are not (later) defined as evil. [Edit 2: WRONG. See below]
- unk_1e6 DFHacked from 256 to 0 did NOT work to convert an evil reanimating region to a "failing reanimation" one, i.e. the new embark, next to a previous one, reanimates as well.
- unk_1e4 DFHacked from 256 to 356 on a good region caused all vegetation to be dead on embark.
- unk_1e4 has so far been seen to be >= 256 only on good and evil regions, not neutral ones. Values greater than 256 has been seen only on evil regions, which is reasonable given the previous observation.
- "dead percentage" > 0 has been seen only together with unk_1e6 = 256.

Current guesses:
- unk_1e4 is actually two fields: one boolean of unknown function, followed by an 8 bit integer one that acts as a dead percentage.
- unk_1e6 is actually two fields where the first one is boolean. The first one would be a "reanimating" flag, and the second one is unknown.

Remains to be investigated:
- Is the "dead percentage" a correct interpretation (i.e. the value range being 0 - 100)?
- What's the missing piece(s) that distinguish an actually reanimating region from one that has the interaction but "doesn't work", apart from the "reanimating" flag?
- What is the function of the unk_1e4 boolean, and is it a boolean?
- What's the function of the second byte of unk_1e6?

Edit: Some basic assumptions seem to have been incorrect. There appears to be two kinds of "reanimation" regions, one non functional one that has a target but no effect, and the second kind that has both a corpse target and a reanimation effect, where only the second kind is functional. Swapping the interactions between a non functional reanimating evil region and a functional one swapped the effect, and assigning a functional reanimating regional interaction to a good region (previously changed to have dead vegetation) caused reanimation to take effect.

Thus, I think the data is:
unk_1e4 byte 1: boolean flag of unknown function
unk_1e4 byte 2: dead percentage (0 - 100) controlling dead vegetation
unk_1e6 byte 1: reanimating. Guides DF to select a regional interaction containing reanimation for the region, but does not control reanimation itself. Thus, it indicates reanimation in an unhacked game (possibly excluding worlds where no reanimating interactions exist).
unk_1e6 byte 2: Boolean flag of unknown function, that may be set only on neutral regions.

I've run the following script on a number of quickly generated medium worlds with good and evil increased to 100 both for small regions, and haven't found any case where anything violating the assumptions has reported.

````function y ()
  for i, region in ipairs (df.global.world.world_data.regions) do
    if region.unk_1e6 == 1 and
       (region.good or region.evil) then
      dfhack.println ("unk_6 = 1 for non neutral region", i, region.unk_1e4, region.unk_1e6, region.good, region.evil)
    end
    
    if region.unk_1e4 > 356 or
       (region.unk_1e4 < 256 and region.unk_1e4 > 100) then
      dfhack.println ("dead percentage out of 0 .. 100 range", i, region.unk_1e4, region.unk_1e6, region.good, region.evil)
    end
    
    if region.unk_1e4 > 0 and region.unk_1e4 ~= 256 and not region.evil then
      dfhack.println ("non zero dead percentage for non evil region", i, region.unk_1e4, region.unk_1e6, region.good, region.evil)
    end
    
    if region.unk_1e4 > 0 and region.unk_1e4 ~= 256 and region.unk_1e6 ~= 256 then
      dfhack.println ("non zero dead percentage when reanimating isn't set", i, region.unk_1e4, region.unk_1e6, region.good, region.evil)
    end
    
    if region.unk_1e6 ~= 0 and region.unk_1e6 ~= 1 and region.unk_1e6 ~= 256 then
      dfhack.println ("unk_1e6 not consisting of two booleans that are never set together", i, region.unk_1e4, region.unk_1e6, region.good, region.evil)
    end
  end
end

y ()``

Edit 2:
The good, evil, and unk_1e4 first byte flag are set during the Good/Evil placement phase during worldgen, while both unk_1e6 flags and the dead percentage flag in unk_1e4 are set during the "other beasts" worldgen phase. It can be noted that df.global.world.interaction_instances are generated during this phase as well, which makes sense.

### PR DESCRIPTION
**Pull request by PatrikLundell**
Closes #237 
Compiled without errors and verified the generated .h file contains the changes.

### CODE REVIEW

File: df.world-data.xml [hunk 1]
```diff
@@ -186,8 +186,10 @@
             <int16_t ref-target='plant_raw' />
         </stl-vector>
 
-        <int16_t name="unk_1e4"/>
-        <int16_t name="unk_1e6"/>
+        <int8_t name="unk_1e4" comment="Probably boolean. Probably optionally set only on good and evil regions"/>
```
Comment (lethosor) [inline df.world-data.xml]: I have a world where `unk_1e4` is set to 0x34

Comment (PatrikLundell) [reply df.world-data.xml]: That means that the second byte contains a dead_percentage of 52. Note that the change SPLITS unk_1e4 into an 8 bit integer and a boolean (unless you refer to my changed definition).

Comment (lethosor) [reply df.world-data.xml]: I'm referring to the first byte of these 4 bytes, which is unk_1e4 after your changes. The second byte is either 0x00 or 0x01 - are you sure it's not a boolean?

Comment (PatrikLundell) [reply df.world-data.xml]: Gah! It means I've screwed up the byte order, so unk_1e4 should be swapped with dead_percentage and unk_1e7 should be swapped with reanimating, i.e.
dead_percentage
unk_1e5
unk_1e6
reanimating

**Comment by lethosor** [comment]
I think the two "unk" fields are booleans as well, after inspecting a fairly large save I have (661 world_region instances) and another, so I changed those to `<bool>`s.

**Pull request merged**